File: testbed_ruby/test/associations/one_to_many_test.rb
require 'test/unit'
require './src-gen-umple/student_j'
require './src-gen-umple/mentor_j'
require './src-gen-umple/program_j'

module CruiseTest
class OneToManyTest < Test::Unit::TestCase

  def test_cannotCreateNullStudent
    assert_raise RuntimeError do #"Unable to create student due to mentor"
      StudentJ.new(99,nil)
    end
  end
  
  def test_CreateMentorWithoutStudent
    m = MentorJ.new("blah")
    assert_equal(0,m.number_of_students)
  end
  
  def test_CreateStudentFromMentor
    m = MentorJ.new("blah")
    s = StudentJ.new(99,m)
  
    assert_equal(99,s.get_number)
    assert_equal(m,s.get_mentor)
    assert_equal(1,m.number_of_students)
    assert_equal(s,m.get_student(0))
  end  
  
  def test_add_studentConstructorInformation
    m = MentorJ.new("blah")
    s = StudentJ.new(10,m)
  
    assert_equal(10,s.get_number)
    assert_equal(m,s.get_mentor)
    assert_equal(1,m.number_of_students)
    assert_equal(s,m.get_student(0))
  end
  
  def test_replaceMentor
    m = MentorJ.new("blah")
    m2 = MentorJ.new("blah2")
  
    s = StudentJ.new(123,m)
    s2 = StudentJ.new(125,m2)
    
    s.set_mentor(m2)
    
    assert_equal(m2,s.get_mentor)
    assert_equal(2,m2.number_of_students)
    assert_equal(s2,m2.get_student(0))
    assert_equal(s,m2.get_student(1))
    
    assert_equal(0,m.number_of_students)
  end 
  
  def test_addToNewMentor
    m = MentorJ.new("blah")
    m2 = MentorJ.new("blah2")
  
    s = StudentJ.new(123,m)
    m2.add_student(s)
    
    assert_equal(m2,s.get_mentor)
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m.number_of_students)
  end  
  
  def test_removeFromExistingMentor
    m = MentorJ.new("blah")
    m2 = MentorJ.new("blah2")
  
    s1 = StudentJ.new(123,m)
    s2 = StudentJ.new(124,m)
    s3 = StudentJ.new(125,m)
    
    assert_equal(3,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    
    m2.add_student(s1)
  
    assert_equal(2,m.number_of_students)
    
    assert_equal(s2,m.get_student(0))
    assert_equal(s3,m.get_student(1))
    
    assert_equal(1,m2.number_of_students)
    assert_equal(s1,m2.get_student(0))
    
    assert_equal(m2,s1.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
  end   
  
  
  def test_cannotSetMentorNull
    m = MentorJ.new("blah")
    s = StudentJ.new(99,m)
    assert_equal(false,s.set_mentor(nil))
    assert_equal(m,s.get_mentor)
    assert_equal(s,m.get_student(0))
  end
  
  def test_set_mentorReplacesExistingMentor
    m = MentorJ.new("blah")
    s = StudentJ.new(99,m)
    
    m2 = MentorJ.new("blah2")
    
    s.set_mentor(m2)
    assert_equal(m2,s.get_mentor)
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m.number_of_students)    
  end  
  
end
end

File: testbed_php/test/associations/OneToManyTest.php
<?php

class OneToManyTest extends UnitTestCase
{

  public function test_cannotCreateNullStudent()
  {
  	$this->expectException(new Exception("Unable to create student due to mentor"));
    new StudentJ(99,null);
  }
  

  public function test_CreateMentorWithoutStudent()
  {
    $m = new MentorJ("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }


  public function test_CreateStudentFromMentor()
  {
    $m = new MentorJ("blah");
    $s = new StudentJ(99,$m);

    $this->assertEqual(99,$s->getNumber());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(1,$m->numberOfStudents());
    $this->assertEqual($s,$m->getStudent(0));
  }  
  

  public function test_addStudentViaConstructorInformation()
  {
    $m = new MentorJ("blah");
    $s = $m->addStudentVia(10);

    $this->assertEqual(10,$s->getNumber());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(1,$m->numberOfStudents());
    $this->assertEqual($s,$m->getStudent(0));
  }
  

  public function test_replaceMentor()
  {
    $m = new MentorJ("blah");
    $m2 = new MentorJ("blah2");

    $s = $m->addStudentVia(123);
    $s2 = $m2->addStudentVia(125);
    
    $s->setMentor($m2);
    
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual(2,$m2->numberOfStudents());
    $this->assertEqual($s2,$m2->getStudent(0));
    $this->assertEqual($s,$m2->getStudent(1));
    
    $this->assertEqual(0,$m->numberOfStudents());
    
  } 
  

  public function test_addToNewMentor()
  {
    $m = new MentorJ("blah");
    $m2 = new MentorJ("blah2");

    $s = $m->addStudentVia(123);
    $m2->addStudent($s);
    
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual($s,$m2->getStudent(0));
    $this->assertEqual(0,$m->numberOfStudents());
  }  
  

  public function test_removeFromExistingMentor()
  {
    $m = new MentorJ("blah");
    $m2 = new MentorJ("blah2");

    $s1 = $m->addStudentVia(123);
    $s2 = $m->addStudentVia(124);
    $s3 = $m->addStudentVia(125);
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    
    $m2->addStudent($s1);

    $this->assertEqual(2,$m->numberOfStudents());
    
    $this->assertEqual($s2,$m->getStudent(0));
    $this->assertEqual($s3,$m->getStudent(1));
    
    $this->assertEqual(1,$m2->numberOfStudents());
    $this->assertEqual($s1,$m2->getStudent(0));
    
    $this->assertEqual($m2,$s1->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
  }   


  public function test_cannotSetMentorNull()
  {
    $m = new MentorJ("blah");
    $s = new StudentJ(99,$m);
    $this->assertEqual(false,$s->setMentor(null));
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($s,$m->getStudent(0));
  }
  

  public function test_setMentorReplacesExistingMentor()
  {
    $m = new MentorJ("blah");
    $s = new StudentJ(99,$m);
    
    $m2 = new MentorJ("blah2");
    
    $s->setMentor($m2);
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual($s,$m2->getStudent(0));
    $this->assertEqual(0,$m->numberOfStudents());    
  }  

  public function test_deleteManyEnd()
  {
    $m = new MentorJ("blah");
    $s = new StudentJ(99,$m);
    $s2 = new StudentJ(98,$m);
    
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual(2,$m->numberOfStudents());

    $s->delete();
    
    $this->assertEqual(null,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual(1,$m->numberOfStudents());
    
    $s2->delete();
    $this->assertEqual(null,$s->getMentor());
    $this->assertEqual(null,$s2->getMentor());
    $this->assertEqual(0,$m->numberOfStudents());
  }
}

?>
